controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go

```diff
@@ -259,6 +262,33 @@ func (in *KubeadmControlPlane) validateCoreDNS() (allErrs field.ErrorList) {
 	return allErrs
 }
 
+func (in *KubeadmControlPlane) validateCoreDNSVersion(prev *KubeadmControlPlane) (allErrs field.ErrorList) {
+	if in.Spec.KubeadmConfigSpec.ClusterConfiguration == nil || prev.Spec.KubeadmConfigSpec.ClusterConfiguration == nil{
+		return allErrs
+	}
+	if prev.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag == "" {
+		return allErrs
+	}
+	dns := &in.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS
+	//return if the type is anything other than empty (default), or CoreDNS.
+	if dns.Type != "" && dns.Type != kubeadmv1.CoreDNS {
+		return allErrs
+	}
+
+	fromVersion := prev.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag
+	if err := migration.ValidUpMigration(fromVersion, dns.ImageTag); err != nil {
```

Comment:
See #2697, there is a different issue where we want to allow upgrading for example from v1.0.6-foobar.1 to v1.0.6-foobar.2, there is some good context in the PR's comments and how we approached this in code.